File: frontend/src/services/notificationService.ts
import api from './api';
import type { ApiResponse, PaginatedData } from '@/types';

export interface NotificationResponse {
  id: number;
  type: string;
  title: string;
  content: string;
  icon?: string;
  color?: string;
  actionUrl?: string;
  actionType?: string;
  actionId?: number;
  isRead: boolean;
  readAt?: string;
  createdAt: string;
  referenceType?: string;
  referenceId?: number;
  actorId?: number;
  actorName?: string;
}

export const notificationService = {
  /**
   * Get all notifications for current user
   */
  getNotifications: async (page: number = 0, size: number = 20): Promise<NotificationResponse[]> => {
    const response = await api.get<ApiResponse<NotificationResponse[]>>('v1/notifications', {
      params: { page, size },
    });
    return response.data.data!;
  },

  /**
   * Mark a notification as read
   */
  markAsRead: async (id: number | string): Promise<void> => {
    await api.put(`v1/notifications/${id}/read`);
  },

  /**
   * Mark all notifications as read
   */
  markAllAsRead: async (): Promise<void> => {
    await api.put('v1/notifications/read-all');
  },

  /**
   * Get unread notification count
   */
  getUnreadCount: async (): Promise<number> => {
    const response = await api.get<ApiResponse<number>>('v1/notifications/unread-count');
    return response.data.data!;
  },
};

export default notificationService;

File: backend/src/main/java/fit/nlu/tmdt/modules/notification/service/NotificationService.java
package fit.nlu.tmdt.modules.notification.service;

import fit.nlu.tmdt.modules.notification.dto.response.NotificationResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;

/**
 * Notification Service Interface
 */
public interface NotificationService {

    /**
     * Get user's notifications
     */
    List<NotificationResponse> getUserNotifications(Long userId);

    Page<NotificationResponse> getUserNotifications(Long userId, Pageable pageable);

    /**
     * Get unread notifications
     */
    List<NotificationResponse> getUnreadNotifications(Long userId);

    /**
     * Get recent notifications
     */
    List<NotificationResponse> getRecentNotifications(Long userId, int limit);

    /**
     * Mark notification as read
     */
    void markAsRead(Long notificationId, Long userId);

    /**
     * Mark all notifications as read
     */
    int markAllAsRead(Long userId);

    /**
     * Get unread count
     */
    long getUnreadCount(Long userId);

    /**
     * Create notification (internal use)
     */
    void createNotification(fit.nlu.tmdt.modules.notification.entity.Notification notification);

    /**
     * Delete old notifications
     */
    int cleanupOldNotifications(Long userId, int daysOld);

    /**
     * Push real-time notification via WebSocket
     */
    void pushNotification(fit.nlu.tmdt.modules.notification.entity.Notification notification);
}